### PR TITLE
Tomcat 5.5+ instrumentation

### PR DESCRIPTION
**Pull request by tylerbenson**
This instrumentation overrides the generic servlet instrumentation and provides a more holistic view of the actual lifecycle of a tomcat request. Particularly important is the better detection of final http status code when an exception is thrown and a different status is returned via an exception handler.

Request/Response objects are always recycled so we use that as the last possible point to ensure the most accurate status code.

(The release notes should indicate that this is a pretty significant change in instrumentation for anyone using Tomcat and to be aware when upgrading. This includes possible impact to duration metrics due to an increased measurement scope.)

### CODE REVIEW

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import java.util.Enumeration;
+import org.apache.coyote.Request;
+
+public class RequestExtractAdapter implements AgentPropagation.ContextVisitor<Request> {
+
+  public static final RequestExtractAdapter GETTER = new RequestExtractAdapter();
+
+  @Override
+  public void forEachKey(Request carrier, AgentPropagation.KeyClassifier classifier) {
+    Enumeration<String> headerNames = carrier.getMimeHeaders().names();
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java:13]: Is this the best `Request` has to offer? I was hoping at this level there would be lower level APIs.

**Review by richardstartin — COMMENTED** [review]
Looks good. I'm surprised how little code is required per container, or how little was saved by treating the servlet API generically.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java (hunk 1 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java:13]: Yes.  You requested similar with the Jetty advice, so I explicitly looked this time and there isn't any way to get an enumeration of the pairs together.

Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java:13]: Specifically here's the API we're dealing with: https://tomcat.apache.org/tomcat-5.5-doc/catalina/docs/api/org/apache/tomcat/util/http/MimeHeaders.html

**Comment by tylerbenson** [comment]
@richardstartin I kind of lucked out in finding an instrumentation point that was so steady over the desired version range so as to not require a bunch of modules like I did for jetty.  Took some digging to find the right level to instrument... the object recycling pattern also helps quite a bit in knowing the last possible point of use.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java [hunk 2]
```diff
@@ -97,10 +102,14 @@ public static void stopSpan(
...
-        if (resp.getStatus() == HttpServletResponse.SC_OK) {
-          // exception is thrown in filter chain, but status code is likely incorrect
-          span.setTag(Tags.HTTP_STATUS, 500);
+        if (!isDispatch) {
+          // We don't want to put the status on the dispatch span.
+          // (It might be wrong/different from the server span with an exception handler.)
+          DECORATE.onResponse(span, resp);
+          if (resp.getStatus() == HttpServletResponse.SC_OK) {
+            // exception is thrown in filter chain, but status code is likely incorrect
+            span.setTag(Tags.HTTP_STATUS, 500);
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: Is this intentional? I thought the idea was to get the correct status?

Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: The original servlet instrumentation still has the same behavior, but this won't be called when tomcat/jetty instrumentation is applied.

Comment (richardstartin) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: I know, but that's not what I'm asking. I thought the driver for going down to the container level was to get accurate status codes, not to lift and shift the servlet instrumentation into container specific instrumentation. If we're still guessing the status code, what's the benefit?

Comment (richardstartin) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: Apologies, I had missed that this is in the servlet 3 advice.

**Review by richardstartin — APPROVED** [review]

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: `Servlet3Advice` is for the generic servlet instrumentation, not the new tomcat/jetty specific instrumentation.

Comment (dougqh) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: I'm thinking this in terms of progress enhancement.  Right now, this code gets us decent broad coverage.  As we fill-in more and more server instrumentations maybe we won't need this anymore.

So I think the right choice is to skip when we know we are using a more feature rich server instrumentation.

File: dd-java-agent/instrumentation/servlet/request-3/request-3.gradle [hunk 3]
```diff
@@ -31,6 +31,7 @@ dependencies {
   testCompile project(':dd-java-agent:instrumentation:jetty-7.0') // See if there's any conflicts.
   testCompile project(':dd-java-agent:instrumentation:jetty-7.6') // See if there's any conflicts.
   testCompile project(':dd-java-agent:instrumentation:jetty-9') // See if there's any conflicts.
+  testCompile project(':dd-java-agent:instrumentation:tomcat-5.5') // See if there's any conflicts.
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/request-3.gradle:34]: Not for this PR, but I think we need to rethink our approach to testing.  Increasingly, I think we should move our tests out into separate parallel components.  And I think those components should use a full tracer rather than just a single instrumentation.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java [hunk 4]
```diff
@@ -15,16 +15,21 @@
...
     this.activated = activated;
     this.span = span;
+    this.isDispatch = isDispatch;
   }
 
   @Override
   public void onComplete(final AsyncEvent event) throws IOException {
     if (activated.compareAndSet(false, true)) {
-      DECORATE.onResponse(span, (HttpServletResponse) event.getSuppliedResponse());
+      if (!isDispatch) {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:30]: The conditional calling of onResponse feels a little messy.  Why do we need that?

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java [hunk 5]
```diff
@@ -45,11 +50,13 @@ public void onTimeout(final AsyncEvent event) throws IOException {
...
   public void onError(final AsyncEvent event) throws IOException {
     if (event.getThrowable() != null && activated.compareAndSet(false, true)) {
-      DECORATE.onResponse(span, (HttpServletResponse) event.getSuppliedResponse());
-      if (((HttpServletResponse) event.getSuppliedResponse()).getStatus()
-          == HttpServletResponse.SC_OK) {
-        // exception is thrown in filter chain, but status code is incorrect
-        span.setTag(Tags.HTTP_STATUS, 500);
+      if (!isDispatch) {
+        DECORATE.onResponse(span, (HttpServletResponse) event.getSuppliedResponse());
+        if (((HttpServletResponse) event.getSuppliedResponse()).getStatus()
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:55]: I know that this shared code in Advice classes is a bit hard to pull out, but this duplication is rather ugly.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy [hunk 6]
```diff
@@ -151,8 +177,8 @@ abstract class AbstractServlet3Test<SERVER, CONTEXT> extends HttpServerTest<SERV
         }
 
         if (endpoint.errored) {
-          "error.msg" { it == null || it == EXCEPTION.body }
-          "error.type" { it == null || it == Exception.name }
+          "error.msg" { it == null || it == EXCEPTION.body || it == CUSTOM_EXCEPTION.body }
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy:180]: The need for this much boolean logic in the test strikes me as messy and hard to follow.

I think this stems from reusing pieces that are too large which requires placing conditional logic in the test.
I think reusing smaller pieces would be much easier to follow.

Right now, this feels hard to trace from what we're actually testing.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy [hunk 7]
```diff
@@ -161,7 +187,7 @@ abstract class AbstractServlet3Test<SERVER, CONTEXT> extends HttpServerTest<SERV
   }
 
   boolean hasResponseSpan(ServerEndpoint endpoint) {
-    return endpoint == REDIRECT || endpoint == ERROR || endpoint == NOT_FOUND
+    return [REDIRECT, ERROR, NOT_FOUND].contains(endpoint)
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy:190]: I probably would have left this as is.  In test code, I think this is fine.  But I'd be opposed elsewhere since this is less efficient.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy [hunk 8]
```diff
@@ -185,7 +211,7 @@ abstract class AbstractServlet3Test<SERVER, CONTEXT> extends HttpServerTest<SERV
           defaultTags()
         }
       }
-    } else if (endpoint == EXCEPTION) {
+    } else if (endpoint == EXCEPTION || endpoint == CUSTOM_EXCEPTION) {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy:214]: Quite minor, but this repetition feels a little messy.  A helper function might help a bit here.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy [hunk 9]
```diff
@@ -44,9 +48,34 @@ abstract class JettyServlet3Test extends AbstractServlet3Test<Server, ServletCon
...
-        Throwable th = (Throwable) request.getAttribute("javax.servlet.error.exception")
-        writer.write(th ? th.message : message)
+        Throwable t = (Throwable) request.getAttribute("javax.servlet.error.exception")
+        def response = ((Request) request).response
+        if (t) {
+          if (t instanceof ServletException) {
+            t = t.rootCause
+          }
+          if (t instanceof InputMismatchException) {
+            response.status = CUSTOM_EXCEPTION.status
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy:73]: Okay, so we're testing this via an ErrorHandle that makes sense to me, but I assume there are other places/APIs that a user could use as well.  I think we need to make sure to check Filter, web.xml config, etc as well.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/TestServlet3.groovy [hunk 10]
```diff
@@ -63,13 +67,23 @@ class TestServlet3 {
...
-        if (resp.class.name.startsWith("org.eclipse.jetty")) {
-          // this line makes Jetty behave like Tomcat and immediately return 500 to the client
-          // otherwise it will continue to repeat the same request until the client times out
-          context.addListener({ resp.status = 500; it.asyncContext.complete() } as AsyncListener)
-        }
+      context.setTimeout(SERVLET_TIMEOUT)
+      if (resp.class.name.startsWith("org.eclipse.jetty")) {
+        // this line makes Jetty behave like Tomcat and immediately return 500 to the client
+        // otherwise it will continue to repeat the same request until the client times out
+        context.addListener(new AsyncListener() {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/TestServlet3.groovy:74]: Okay, I worry a bit about forcing the server into non-standard way of functioning.
But from the comment, it sounds like it makes sense in this instance.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/TestServlet3.groovy [hunk 11]
```diff
@@ -143,6 +154,8 @@ class TestServlet3 {
               break
             case EXCEPTION:
               throw new Exception(endpoint.body)
+            case CUSTOM_EXCEPTION:
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/TestServlet3.groovy:157]: It looks like these factories are repeated in multiple places.  Is there not a better way to share this code?

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy [hunk 12]
```diff
@@ -84,6 +81,7 @@ abstract class AbstractServlet3Test<SERVER, CONTEXT> extends HttpServerTest<SERV
     addServlet(context, "/dispatch" + REDIRECT.path, dispatchServlet)
```
Comment (dougqh) [inline dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy:81]: I don't think we need to fix this now, but...

It seems like it would have been better to create a number of different servlet instances rather than push so much conditional logic into the test servlet.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java (hunk 1 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java:13]: Yes, I'm surprised the new version isn't better, but so it is.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/ResponseInstrumentation.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import org.apache.catalina.connector.CoyoteAdapter;
+import org.apache.catalina.connector.Request;
+import org.apache.catalina.connector.Response;
+
+@AutoService(Instrumenter.class)
+public final class ResponseInstrumentation extends Instrumenter.Tracing {
+
+  public ResponseInstrumentation() {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/ResponseInstrumentation.java:24]: Definitely not part of this PR, but I'm wondering if we need to tackle some of the revised instrumentation concepts.
It seems like this tomcat instrumentation is really tomcat that depends on servlet.
It also seems like there are really multiple instrumentations that make up the single logical tomcat instrumentation.

Those instrumentations should all be activated as a single atom.

But we can come back to that _after_ we've fixed addressed the more user visible aspect of the server integrations.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+    public static void closeScope(@Advice.Enter final AgentScope scope) {
+      scope.close();
+    }
+
+    private void muzzleCheck(CoyoteAdapter adapter, Request request, Response response)
+        throws Exception {
+      adapter.service(null, null);
```
Comment (dougqh) [inline dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java:101]: Again not for this PR, but yes, the need to use muzzleCheck definitely suggests that we need to work on atomically activating a family of instrumentations.

File: dd-java-agent/instrumentation/tomcat-5.5/src/test/groovy/TomcatServletTest.groovy [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+import static datadog.trace.agent.test.base.HttpServerTest.ServerEndpoint.CUSTOM_EXCEPTION
+import static datadog.trace.agent.test.base.HttpServerTest.ServerEndpoint.EXCEPTION
+import static datadog.trace.agent.test.base.HttpServerTest.ServerEndpoint.SUCCESS
+import static datadog.trace.agent.test.base.HttpServerTest.ServerEndpoint.TIMEOUT
+import static datadog.trace.agent.test.base.HttpServerTest.ServerEndpoint.TIMEOUT_ERROR
+import static org.junit.Assume.assumeTrue
+
+@Unroll
+class TomcatServletTest extends AbstractServletTest<Embedded, Context> {
```
Comment (dougqh) [inline dd-java-agent/instrumentation/tomcat-5.5/src/test/groovy/TomcatServletTest.groovy:29]: It feels like we might have gone slightly overboard with the generics here.
I feel like having an Embedded member var in the each test might have been easier to follow.

I don't feel strongly about this.  Just food for thought.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java (hunk 14 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java:101]: This wasn't strictly necessary, but muzzle doesn't check the availability of the method we are instrumenting (unless it's called by the advice somewhere, which it usually isn't).  I added this call to ensure that the range match takes that into consideration.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java (hunk 4 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:30]: When we do an async dispatch, there are times when the status code can be set differently from the root server span. (Specifically the wrong one being set by the servlet instrumentation and the correct one being set by the tomcat/jetty instrumentation.) I opted to not report status code on this span rather than risk it being incorrect.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java (hunk 5 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:55]: This particular class isn't an advice class, so we can refactor it better.  Can you be more specific about what duplication you'd like removed?

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy (hunk 7 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy:190]: I understand this is less performant, but it seemed a worthy tradeoff for consistency.  There are other places in this PR where the list is 5 elements long, so the improved readability seemed better to me.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy (hunk 9 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy:73]: Ok, I can add those test cases.  Do you want it added to this PR or can it be a separate PR?

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/ResponseInstrumentation.java (hunk 13 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/ResponseInstrumentation.java:24]: Not arguing your point, but just to clarify... from what I was able to see Tomcat isn't really usable independent from Serlvet, unlike Jetty, which doesn't require it.

File: dd-java-agent/instrumentation/tomcat-5.5/src/test/groovy/TomcatServletTest.groovy (hunk 15 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/test/groovy/TomcatServletTest.groovy:29]: Here the generics are certainly not necessary.  This is purely an artifact of copy/paste from the original `AbstractServletTest`

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy (hunk 12 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy:81]: This might be worth adding to the backlog to refactor.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java (hunk 1 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java:13]: This didn't look right and what's been said here about the API isn't accurate. The headers can be iterated over without a quadratic loop using these methods

https://tomcat.apache.org/tomcat-5.5-doc/catalina/docs/api/org/apache/tomcat/util/http/MimeHeaders.html#getName(int)
 https://tomcat.apache.org/tomcat-5.5-doc/catalina/docs/api/org/apache/tomcat/util/http/MimeHeaders.html#getValue(int)

```java
MimeHeaders headers = carrier.getMimeHeaders();
for (int i = 0; i < headers.size(); ++i) {
     MessageBytes header = headers.getName(i);
     MessageBytes value = headers.getValue(i);
     if (!classifier.accept(header.toString(), value.toString())) {
        return;
      }
}
```

`MessageBytes` has a lot of [support](https://tomcat.apache.org/tomcat-5.5-doc/catalina/docs/api/org/apache/tomcat/util/buf/MessageBytes.html) for the obvious inefficiency above - converting headers and values we don't even know we're interested in yet into strings - but there was pushback against there being a little bit of code duplication when using lower level APIs. What's above is still more efficient than getting an enumeration and doing a linear search for each value in the enumeration.

**Review by richardstartin — CHANGES_REQUESTED** [review]
Please make use of `MimeHeaders`' support for linear time context extraction

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java (hunk 1 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java:13]: Sensibly, the string is actually cached anyway, and may have been initialised already, or would be initialised later without the tracer existing. https://github.com/apache/tomcat/blob/681f2afccc2f22ff5fc3d80ad7e77dbeecd083b2/java/org/apache/tomcat/util/buf/MessageBytes.java#L163L166

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java (hunk 2 above)
Comment (richardstartin) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java:111]: Nit: this is boxing an `int` outside of the default range of the `Integer` cache.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java [hunk 16]
```diff
@@ -45,11 +50,13 @@ public void onTimeout(final AsyncEvent event) throws IOException {
...
+            == HttpServletResponse.SC_OK) {
+          // exception is thrown in filter chain, but status code is incorrect
+          span.setTag(Tags.HTTP_STATUS, 500);
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:58]: Boxing

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy (hunk 6 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy]: Good call out.  I cleaned this up some and discovered the exception wasn't being set in some cases I was expecting it, so I updated the decorators to capture it properly.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy (hunk 8 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/AbstractServlet3Test.groovy]: replaced with `endpoint.throwsException`

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java (hunk 1 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/RequestExtractAdapter.java]: @richardstartin good call out.  I made that change and was able to find a similar API in jetty. Thanks for the idea.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java (hunk 2 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java]: This was being done in several other places too, so I applied the cache lookup everywhere.

Comment (richardstartin) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java]: Previously, we had this logic in `RequestDispatcherDecorator`

```java
        int status = (int) STATUS_CODE_METHOD.invokeExact((HttpServletResponse) response);
        if (status > UNSET_STATUS) {
          if (throwable != null && status == HttpServletResponse.SC_OK) {
            span.setTag(Tags.HTTP_STATUS, SERVER_ERROR);
            span.setError(true);
          } else {
            span.setTag(Tags.HTTP_STATUS, HTTP_STATUSES.get(status));
            if (SERVER_ERROR_STATUSES.get(status)) {
              span.setError(true);
            }
          }
```

and we had the following

```java
private static final Integer SERVER_ERROR = 500;
```

so we could avoid boxing and avoid the lookup (which is cheap but not free), but it looks like there are indeed a few other places where this idiom was in place. Could we use the constant where the value is known in this PR please (otherwise someone else will have to take the time to apply the change after the branch is merged)?

Comment (tylerbenson) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/Servlet3Advice.java]: good catch.  I've replaced that with a static final in `HttpServerDecorator` rather than using the cache directly.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java (hunk 4 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:31]: Hmm, that makes me wonder if there isn't a better place to instrument.  At the very least, I think this needs a lot more commenting as to what is going on here and why.

File: dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java (hunk 5 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/servlet/request-3/src/main/java/datadog/trace/instrumentation/servlet3/TagSettingAsyncListener.java:56]: This is repeating the same exception/500 status code handling in multiple places.  Also, this location isn't using the caching as mentioned in another comment.

File: dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy (hunk 9 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/servlet/request-3/src/test/groovy/JettyServlet3Test.groovy:73]: I'm inclined to say this PR because right now I don't have a lot of confidence that the end user is actually fixed.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/ResponseInstrumentation.java (hunk 13 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/ResponseInstrumentation.java:24]: Yes, I understand that Tomcat requires Servlet -- that's what I said originally.
However, Servlet is intended to be used without Tomcat.

As said, not for this PR, but I think we need to rethink instrumentation and test dependencies.

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java (hunk 14 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java:101]: Now, I'm more confused.  Why not just build that functionality into muzzle?

File: dd-java-agent/instrumentation/tomcat-5.5/src/test/groovy/TomcatServletTest.groovy (hunk 15 above)
Comment (dougqh) [reply dd-java-agent/instrumentation/tomcat-5.5/src/test/groovy/TomcatServletTest.groovy:29]: Okay.  I have noticed a tendency to overuse generics throughout the code base that I find more constraining than helpful.

**Review by richardstartin — APPROVED** [review]
My concerns have been addressed.

**Review by bantonsson — APPROVED** [review]

File: dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java (hunk 14 above)
Comment (tylerbenson) [reply dd-java-agent/instrumentation/tomcat-5.5/src/main/java/datadog/trace/instrumentation/tomcat/TomcatServerInstrumentation.java:101]: That's certainly something that could be done, but I don't have a good idea for how to do that.  I think the way matching works makes this difficult.

**Comment by tylerbenson** [comment]
I think I've resolved all the concerns...
